yellow-paper/docs/rollup-circuits/index.md

```diff
@@ -0,0 +1,429 @@
+
+### Higher-level tasks
+Before looking at the circuits individually, it can however be a good idea to recall the reason we had them in the first place. For this, we are especially interested in the tasks that goes across multiple circuits and proofs.
+
+#### State consistency
+While the individual kernels are validated on their own, they might rely on state changes earlier in the block. For the block to be correctly validated, this means that when validating kernel $n$, it must be executed on top of the state after all kernels $<n$ have been applied. For example, when kernel $3$ is executed, it must be executed on top of the state after kernel $0$, $1$ and $2$ has been applied. If this is not the case, the kernel proof might be valid, but the state changes invalid which could lead to double spends. 
+
+It is therefore of the highest importance that the circuits ensure that the state is progressed correctly across circuit types and proofs. Logically, taking a few of the kernels from the above should be executed/proven as butchered below, $k_n$ applied on top of the state that applied $k_{n-1}$
+
+```mermaid
+graph LR
+    SM[State Machine]
+    S0((State 0))
+    K0((Kernel 0))
+    S1((State 1))
+
+    S0 --> SM
+    K0 --> SM
+    SM --> S1
+
+
+    SM_2[State Machine]
+    K1((Kernel 1))
+    S2((State 2))
+
+    S1 --> SM_2
+    K1 --> SM_2
+    SM_2 --> S2
+
+    style K0 fill:#1976D2;
+    style K1 fill:#1976D2;
+```
+
+#### State availability
+To ensure that state is made available, we could rely on the full block body as public inputs, but this would be very expensive. Instead we rely on a commitment to the body (the `ContentHash`) which we build upon throughout proof generation as needed.
+
+To check that this body is published a node can reconstruct the `ContentHash` from available data. Since we define finality as the point where the block is validated and included in the state of the [validating light node](./../contracts/index.md), we can define "available" at the level of this node, e.g., if the validating light node can reconstruct the commitment then it is available.
+
+Since we strive to minimize the compute requirements to prove blocks, we amortize the commitment cost across the full three. We can do so by building merkle trees of partial "commitments" that are then finished at the root. Below, we outline the `TxsHash` merkle tree that is based on the `TxEffect`s and a `OutHash` which is based on the `l2_to_l1_msgs` (cross-chain messages) for each transaction. While the `TxsHash` implicitly include the `l2_to_l1_msgs` we construct it separately since the `l2_to_l1_msgs` must be known to the contract directly and not just proven available. This is not a concern when using calldata as the data layer, but is a concern when using alternative data layers such as [Celestia](https://celestia.org/) or [Blobs](https://eips.ethereum.org/EIPS/eip-4844).
```

Comment:
```suggestion
Since we strive to minimize the compute requirements to prove blocks, we amortize the commitment cost across the full tree. We can do so by building merkle trees of partial "commitments" that are then finished at the root. Below, we outline the `TxsHash` merkle tree that is based on the `TxEffect`s and a `OutHash` which is based on the `l2_to_l1_msgs` (cross-chain messages) for each transaction. While the `TxsHash` implicitly include the `l2_to_l1_msgs` we construct it separately since the `l2_to_l1_msgs` must be known to the contract directly and not just proven available. This is not a concern when using calldata as the data layer, but is a concern when using alternative data layers such as [Celestia](https://celestia.org/) or [Blobs](https://eips.ethereum.org/EIPS/eip-4844).
```